src/model/rpc.js

```diff
@@ -104,14 +154,93 @@ module.exports = class BtpRpc extends EventEmitter {
     }
   }
 
+  // authentication handling must be done in the RPC, because it's done
+  // on a per-socket basis rather than a per-plugin basis.
+  async _handleAuth (socketIndex, { type, requestId, data }) {
+    const socketData = this._sockets[socketIndex]
+    this.debug('authenticating socket #' + socketIndex)
+
+    if (type !== btpPacket.TYPE_MESSAGE) {
+      this.debug(`responding to invalid auth request: ${JSON.stringify(data)}`)
+      await _send(socketData.socket, btpPacket.serializeError({
+        code: 'F01',
+        name: 'InvalidFieldsError',
+        triggeredAt: new Date(),
+        data: JSON.stringify({ message: 'invalid method on unauthenticated socket' })
+      }, requestId, []))
+      return
+    }
+
+    if (data.protocolData[0].protocolName !== 'auth') {
+      this.debug(`responding to invalid auth request: ${JSON.stringify(data)}`)
+      await _send(socketData.socket, btpPacket.serializeError({
+        code: 'F01',
+        name: 'InvalidFieldsError',
+        triggeredAt: new Date(),
+        data: JSON.stringify({ message: 'auth must be primary protocol on unauthenticated message' })
+      }, requestId, []))
+      return
+    }
+
+    const [ authToken ] = data.protocolData.filter(p => p.protocolName === 'auth_token')
+    if (!authToken) {
+      this.debug(`responding to invalid auth request: ${JSON.stringify(data)}`)
+      await _send(socketData.socket, btpPacket.serializeError({
+        code: 'F01',
+        name: 'InvalidFieldsError',
+        triggeredAt: new Date(),
+        data: JSON.stringify({ message: 'missing "auth_token" secondary protocol' })
+      }, requestId, []))
+      return
+    }
```

Comment:
There is a lot of duplicated code from line 163 to 207. How about refactoring that into a new function?